### PR TITLE
Lowlevel Refactor

### PR DESCRIPTION
**Pull request by samuelsadok**
Summary of the most important changes:

 - Replace MotorControl/board_config_v3.h by Board/v3/board.cpp (hardware specific variable instantiations and functions) and Board/v3/Inc/board.h (hardware specific declarations and defines)
 - Move the scaffolding of the boot sequence from Board/v3/Src/main.c and Board/v3/Src/freertos.c to MotorControl/main.cpp (including the `main()` entry point).
 - Statically allocate Axis objects
 - Overhaul NVM code to allow for cleaner code organization and facilitate (future) dynamically sized config items.
 - Introduce SPI arbiter to allow multiple components to share the same SPI in a thread-safe way.
 - Introduce explicit GPIO modes in order to avoid unclear and fragile priorities that were a side effect of the init order.
 - Treat A/B/Z encoder I/O and on-board CAN lines as GPIO9...GPIO16. This opens up the possibility to disable the incremental encoders and use their pins for other purposes (e.g. I2C).

HWIL test log:
 171dab31bd3c241c679119813a6b88fd591827bc: all tests pass
 4b0a6fa478d7ba7b7bb318f110c863887c696e43: all tests pass (`analog_input_test.py`, `calibration_test.py`, `can_test.py`, `closed_loop_test.py`, `encoder_test.py`, `fibre_test.py`, `integration_test.py`, `nvm_test.py`, `pwm_input_test.py`, `step_dir_test.py`, `uart_ascii_test.py`)

### CODE REVIEW

**Comment by jhurliman** [comment]
Is this ready for testing? I'm wondering if it will solve our issues of the AMT232B and DRV8301 fighting with each other.

**Comment by jhurliman** [comment]
I tried merging this branch into devel and building/flashing it. The device gets stuck in the init sequence just like https://github.com/madcowswe/ODrive/issues/420 except now it happens with or without pull-up resistors.

**Comment by samuelsadok** [comment]
Yes this is ready for testing.
However I wouldn't expect that it solves your issue; I don't think there's a software-caused SPI collision on `devel` because of the order in which things are initialized. I think we need to get to understand your issue separately over at #420.

File: Firmware/Board/v3/board.cpp [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+        encoders[1], // encoder
+        sensorless_estimators[1], // sensorless_estimator
+        controllers[1], // controller
+        fet_thermistors[1], // fet_thermistor
+        motor_thermistors[1], // motor_thermistor
+        motors[1], // motor
+        trap[1], // trap
+        endstops[2], endstops[3], // min_endstop, max_endstop
+    },
+}};
```
Comment (madcowswe) [inline Firmware/Board/v3/board.cpp:139]: is it possible to do the struct initialization by name thing like `.stuff = bla` ?

File: Firmware/Board/v3/board.cpp [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+    {GPIOC, GPIO_PIN_9}, // ENC0_Z
+    {GPIOB, GPIO_PIN_6}, // ENC1_A
+    {GPIOB, GPIO_PIN_7}, // ENC1_B
+    {GPIOC, GPIO_PIN_15}, // ENC1_Z
+    {GPIOB, GPIO_PIN_8}, // CAN_R
+    {GPIOB, GPIO_PIN_9}, // CAN_D
+};
+#else
+#error "unknown GPIOs"
+#endif
```
Comment (madcowswe) [inline Firmware/Board/v3/board.cpp]: Please add a comment next to each one so it's easy for human to see which gpio is what at a glance.

**Review by madcowswe — COMMENTED** [review]
I don't think I will be able to make time to review the rest, just left two basic comments.
Just make sure that the test coverage is decent, then you can just merge it in.

File: Firmware/Board/v3/board.cpp (hunk 1 above)
Comment (samuelsadok) [reply Firmware/Board/v3/board.cpp:139]: Unfortunately no because these are initializer lists that get applied as arguments to the `Axis` constructor.

**Pull request merged**